### ISSUE TITLE
Feature: Boost volume for songs that are recorded at low volume with high momentary peaks

### ISSUE THREAD

**Issue by mpogue2** [mpogue2/SquareDesk#1220]
This could be an icon-enabled feature, like Loop On and Snap On.

I have a couple of songs that still do not have enough volume, even with normalization on, because they have maybe a single peak that is close to 1.0, and the rest of the song is around 0.5, so it's way low.

Rather than allowing the volume control to go past 1.0/MAX, I'm thinking that a persistent +6dB boost would do, enabled by clicking on the icon.  Could even just be the characters "+6" in a legible font on a button.

I already tried a Compressor, but that didn't work well.

**Comment by pkbott** [comment]
I have quite a few songs like this as well, where this feature might be
very useful.

**Comment by Gero5** [comment]
Same thing here - as discussed before. Especially with old vinyl recordings.

> Rather than allowing the volume control to go past 1.0/MAX

Bummer. So far that was my favorite idea to solve that. Or a separate boost knob or slider.
I am not convinced that a fixed +6dB ist suitable for all.

**Comment by Gero5** [comment]
>because they have maybe a single peak that is close to 1.0, and the rest of the song is around 0.5

Right. So far the Normalization option is quite unsatisfying to solve this problem.

>I already tried a Compressor, but that didn't work well.

For getting more punch to a track in DAW mixing I prefer a limiter over compressors in the first. It does the job without the pumping effect that compressors have. 

One of the most versatile tool for mixing is the Oxford Inflator.  It is not open source but you can get a demo and see how it works on your low volume songs in a DAW to see if the effect gets in the right direction.
https://sonnox.com/products/oxford-inflator

It does exactly what you mention above - boosting the loudness, cutting away singular peaks without a negative  influence on the overall sound. And, of course, the limiter avoids all overdrive problems that you might get with a simple linear boost. 

Maybe we could find something similar for SquareDesk?

**Comment by mpogue2** [comment]
Audio Hijack makes testing something like LoudMax far easier, but it's not free.
So, to play around with the LoudMax solution (using only FREE AS IN BEER components):

BLACKHOLE:
- download and install Blackhole (FREE 2ch version: https://existential.audio/blackhole/)
    - set your system audio OUTPUT to "Blackhole 2ch" to send all audio to Blackhole

LOUDMAX:
- install LoudMax AU plugin (FREE: https://loudmax.blogspot.com)
- download and install Element (FREE: https://kushview.net/element/download/)
    - run Element's Plugin Manager, and click "Scan" to find all AU's
    - create a graph that goes from Blackhole 2ch to LoudMax and then to your Speakers, like in the diagram below
    - leave Element up and running

SQUAREDESK:
- start SquareDesk
    - check in lower right corner (status bar) that output is "Blackhole 2ch"
    - load a song, and start playing it
- dynamically change LoudMax settings to hear what it does to the song

WHEN DONE:
- remember to change your system audio back to "MacBook Air Speakers" (or equivalent)
- save your Element session somewhere, so you can just load it back later
- shutdown Element

![image](https://github.com/user-attachments/assets/e5b963e1-4594-44ec-9f03-d3c926f749ec)


<img width="689" alt="image" src="https://github.com/user-attachments/assets/2c7696b9-cda7-4542-8ed1-3ebf7d2ce15d">

I tested this on an M1 MacBook Air, and it worked flawlessly.

**Comment by mpogue2** [comment]
And, by the way, I did like the way LoudMax did "compression" without doing compression... :-)

**Comment by pkbott** [comment]
Just my 2 cents but in my Christmas light show world I was told to use MP3 gain to do something similar. I personally used the program and like it. Just this year it also became part of another part of my Christmas lights software FPP which is used to run it. Whoever customized our free as in beer money software added mp3 gain to it. Might want to check it out?PaulReady to try something FUN, UNIQUE and ***@***.*** Square Dance Group:  https://www.facebook.com/groups/squaredanceforfun/On Nov 30, 2024, at 3:24 AM, Mike Pogue ***@***.***> wrote:﻿
And, by the way, I did like the way LoudMax did "compression" without doing compression... :-)

**Comment by mpogue2** [comment]
@pkbott SquareDesk used to have mp3gain built-in.  It didn't work very well for leveling volume, so I removed it!

**Comment by Gero5** [comment]
> Oxford Inflator sounds a bit "magical" to me...

As a matter of fact it is. And once used you get easily addicted to it. But, anyhow, I did not expect it to be the right choice for Square Desk because then an universal VST interface has to be implemented which is a bit over the top IMO.  

LOUDMAX:
Tried it in my DAW and from what I hear I think it is very suitable for the given task. The resulting sound is pleasing to me.
And it is easy to use: Set the OUT slider to -1 dB for all songs and only the THRESH value needs to be adjusted.

Would you set the THRESH value to a static number? The high sophisticated solution would rather be to calculate the RMS loudness or LUFS value for the song  and automatically adjust the THRESH value according to that result.

**Comment by mpogue2** [comment]
A universal VST interface for a single VST might not be too bad to implement.  There are several open source VST "hosts" out there that we could draw upon.  A full stack VST implementation would be a lot more work, I think.  

IIRC, the hard part about supporting a single VST is handling the GUI.  Maybe Element's source code can help us here, since it's an example of how it might be done.

LOUDMAX: We could set it to a single value for all songs, or (more complex) we could set it to a different value for each song.  Calculating the RMS or LUFS value is similar to what @pkbott was suggesting -- mp3gain does essentially that.  Used to do that, but the result didn't sound good.  Perhaps "mp3gain + LoudMax" would work better?  Would it work better than just setting a single value for all songs?  That would require some research.  

I kinda like the idea of supporting a single VST, because then somebody might choose the awesome-but-complex TDR Nova plugin (https://www.tokyodawn.net/tdr-nova/) instead of the simpler-but-sounds-pretty-good LoudMax.  

But, as you suggest, supporting even one VST might be challenging.

**Comment by mpogue2** [comment]
By the way, in my testing, when sliding the top slider to the left, if the song is a good volume level, LoudMax doesn't seem to do anything at all.  It only seems to kick in, when the volume of the song is low.  Because of this, maybe we would not need a different setting for each song, and a single setting would work well enough?

**Comment by Gero5** [comment]
> when sliding the top slider to the left, if the song is a good volume level, LoudMax doesn't seem to do anything at all.

Due to my testing results I would phrase it like this: It does not harm the music too much. But there is a lot more compression involved. You can hear that if you have music parts at a lower loudness level than the rest.

Here are the the first 67 seconds of California Dreaming that starts with a low level guitar intro. The music is a recorded version from the vinyl record and has a lot of sparkling cracks in it that are typical for vinyls that have met a Hilton turntable too often. And as some cracks are at 0 dB so normalization does have no effect. 
![Bildschirmfoto 2024-12-01 um 17 13 42](https://github.com/user-attachments/assets/f2bf6790-5ba1-438f-90bd-86a867f128cd)

This is the same song with LOUDMAX at -10 / -1 which I would say is the optimal pick. Note that the guitar intro stays in the lower dynamic.
![Bildschirmfoto 2024-12-01 um 17 20 18](https://github.com/user-attachments/assets/047a1b8e-1692-439e-a703-1bd679b6d07a)

Same song with LOUDMAX at Thresh:Left (-30) / -1. Look at the start of the song. The lower loudness of the intro is completely gone. However, after the intro there is no hearable difference in the loudness between -30 and -10 to me. Although you can very well see it in the sound envelope that there is more compression.
![Bildschirmfoto 2024-12-01 um 17 25 23](https://github.com/user-attachments/assets/7f5470be-f1c8-464c-8bc0-1254e4933089)

All samples: [California Dreaming LOUDMAX.zip](https://github.com/user-attachments/files/17969643/California.Dreaming.LOUDMAX.zip)

Anyhow I can well live with the result of both values. Maybe the LOUDMAX threshold could be adjusted in the preferences? I found out that -9 ... -12 is a good pick for the vinyl recordings I have. But mine were all normalized after recording. 

Does LOUDMAX add distortion? Yes a bit but very little. You can hear it when you carefully listen to the lead instrument that there are some overtones that are not in the original recording. Limiting san Compression is not possible without that effect. For the results given I think we can well live with that for Square Dancing purposes. 
Maybe you want to test if it has negative effects that add up when speed an pitch adjustments are made?

**Comment by mpogue2** [comment]
NOTES:
- JUCE seems to compile OK on M1, but it doesn't run the examples I built.
- OTOH, ProJucer seems to generate code that both builds and runs in XCode
- some hints for getting started with JUCE and enumerating plugins (https://forum.juce.com/t/hosting-plugins/44560/4)

**Comment by mpogue2** [comment]
Moving the progress reports over to #1258.

**Comment by mpogue2** [comment]
Marking this one "Ready for Final Check", because LoudMax support is in now, and it has been stable.
